Fix trace call for inner reverts

ngotchac opened the pull request:
The native `TraceCall` tracer wasn't capturing the output/revert reason if last inner calls were reverting. This fixes this, and adds a new test based on this dummy contract:
```
contract Revertor {
    function run() public pure {
        require(2 > 3, "This called failed");
    }
}

contract Contract {
    constructor() {
        Revertor r = new Revertor();
        r.run();
    }
}
```

I had to tweak a bit the tests setup because it couldn't detect the transaction I sent using `geth --dev`

### Backwards-compatibility note

In a call frame of type CREATE, the field `to` will contain normally the address of the created address. In case of call failure the `to` field contained inconsistent values depending on context. If call frame was the top-level call, then the `to` field still contained the address, but for nested calls the `to` field was cleared to the `0x000...00` address. This inconsistency has been fixed. I.e. a failing top-level call frame of type CREATE will have its `to` field set to `0x000...00`.

holiman commented:
Nice catch! 
It would be good if we can also commit the solidity contracts _somewhere_, if not in the json, then maybe in the go-code somewhere, or an accompanying readme. Not sure, up for suggestions.

holiman commented on eth/tracers/internal/tracetest/testdata/call_tracer/inner_revert_reason.json:
```diff
@@ -0,0 +1,86 @@
+    "from": "0x3623191d4ccfbbdf09e8ebf6382a1f8257417bc1",
+    "gas": "0x2cd774",
+    "gasUsed": "0x16644",
+    "to": "0x0000000000000000000000000000000000000000",
```
This looks strange, to me

ngotchac replied:
So that's because it's a contract creation, but I guess there shouldn't be any `to` field then?

holiman commented on eth/tracers/native/call.go:
```diff
@@ -56,6 +56,29 @@ func (f callFrame) TypeString() string {
+		return
+	}
+	f.Error = err.Error()
+	if f.Type == vm.CREATE || f.Type == vm.CREATE2 {
+		f.To = common.Address{}
+	}
```
Hm, with `omitempty` set, from https://github.com/ethereum/go-ethereum/pull/25637/files, the idea is that `0x000..` should not be present in the output json. So it's basically `null`. Which is a bit wonky too, but.. Still I wonder why we have this clause there.

holiman replied:
I think the oriignal difference between outer scope vs inner scope, which was introduced here: https://github.com/ethereum/go-ethereum/commit/8d7e6062ece1550936065aa8a8f8c7aacc584810#diff-c7b3691786d65b22103e27eca0ca1e6650ddf78874b7326bf7afcc30bf6d2b2fR118 -- is because it originally needed to look up the call in a `t.callstack`. And the outermost scope didn't have preceding entry in the callstack. 

As for why the to was set to empty, https://github.com/ethereum/go-ethereum/commit/8d7e6062ece1550936065aa8a8f8c7aacc584810#diff-c7b3691786d65b22103e27eca0ca1e6650ddf78874b7326bf7afcc30bf6d2b2fR131 , I'm not completely sure.

ngotchac replied:
So here I just kept what was there for the `CaptureExit` case. Regarding a "null" `to` field, maybe we can have a `To *common.Address` (so a pointer) with `omitempty`, so that we can have trully null values, different from just the `0x00..00` address. WDYT?

ngotchac commented:
> It would be good if we can also commit the solidity contracts _somewhere_, if not in the json, then maybe in the go-code somewhere, or an accompanying readme.

Yeah I wanted to put the contract code somewhere but didn't really find where. I can add a `inner_revert_reason.md` next to the `inner_revert_reason.json` file for this maybe?

holiman commented:
I think there are two things that needs to be decided. 

1. For failed CREATE, should we clear the address-which-wasn't-created ? 
1.a Clear it for non-toplevel scope (current master-behaviour)
1.b. Clear it for all scopes (this PR) 
1.c Don't clear it

I prefer 1.b over 1.a, for consistency. And I suspect the reason we clear it is to make it easy for a caller to tell whether a contract was created or not. So I'm leaning towards `1.b` --  _mostly_ keeping the current behaviour, but making it more consistent. @s1na WDYT? 

Then there's a second decision to make, which is how to handle `omitempty`. If someone sends an ether to the `0x00...00` address, then the address should be shown, as a regular address. Not hidden due to `omitempty` rules. Therefore, we may need to redefine something as a `*common.Address`.  I don't think we need to solve this second problem in this PR.

holiman commented:
> I can add a inner_revert_reason.md next to the inner_revert_reason.json file for this maybe?

Sounds good to me!

holiman commented:
Ah, actually.. 
```golang
func TestFooBar(t *testing.T) {
	type test struct {
		Address common.Address `json:"to,omitempty"`
		BytesA  hexutil.Bytes  `json:"bytesA,omitempty"`
		BytesB  hexutil.Bytes  `json:"bytesB,omitempty"`
		String  string         `json:"string,omitempty"`
	}

	var enc test
	enc.Address = common.Address{}
	enc.BytesA = make([]byte, 0, 0)
	enc.BytesB = nil
	enc.String = ""
	outp, err := json.MarshalIndent(&enc, "", "  ")
	if err != nil {
		t.Fatal(err)
	}
	t.Logf("enc: %v\n", string(outp))
}

```
Output: 
```
=== RUN   TestFooBar
    calltrace_test.go:403: enc: {
          "to": "0x0000000000000000000000000000000000000000"
        }
--- PASS: TestFooBar (0.00s)
```
So, the `omitempty` is useless on `common.Address`. So the problem is not `calls to zero looks like contract-creation`, but rather `contract-creation looks like calls to zero`.

ngotchac commented:
@holiman I added a description in a Markdown for the added test (with the contract code), I updated the `callFrame` definition for `To` to have a pointer, so it's `nil` when no contract was created, and I updated the tests' results accordingly. Let me know if this works for you!

holiman commented:
Ping @s1na re https://github.com/ethereum/go-ethereum/pull/25971#issuecomment-1277479389

s1na commented:
So when I ported the tracer to Go I merely adopted this behavior which I also suspect is to make it easier to detect when creation failed. But you're right that it's inconsistent now. My preference would be 1c (not clearing it for any scope). The reasoning is that the error field can be checked to see whether creation failed or not. And users will still be able to see the potential address. E.g. in case of collision they can check what exists at that address.

ngotchac commented:
I think you're right @s1na ; I updated the PR so it's doesn't change the `To` field in case of errors, and updated the test data accordingly.

s1na commented:
I'm sorry for the extra hassle but can you please revert the `To` clearing logic? we decided to do that for logs in https://github.com/ethereum/go-ethereum/pull/25991#issuecomment-1293193252, and it would be good to stay consistent on this behavior. After all created contract and logs are both artefacts that are discarded when a call fails.

Also in the meanwhile #26048 was merged which has some merge conflicts with this PR. The gist of the change there is gasUsed field for the top-level call is now computed in `TxEnd` instead of `End` hook.

ngotchac commented:
No worries, all good! I'll change it back

> please revert the To clearing logic

Would it be better to have `To common.Address`, thus set to `0x00..00` or `To *common.Address` set to `nil`/inexistant in the marhsalled JSON?

ngotchac commented:
@s1na Done with cleared `To` set to `0x00..00`

holiman commented:
> and it would be good to stay consistent on this behavior. 

@s1na good call -- also because it's closer to the current behaviour, so less chance of breaking things.

s1na reviewed (approved):
LGTM thanks!
